src/logging/wal_recovery.cpp

```diff
@@ -0,0 +1,589 @@
+    key_schema->SetIndexedColumns(key_attrs);
+
+    index::IndexMetadata *index_metadata = new index::IndexMetadata(
+            tup->GetValue(1).ToString(), tup->GetValue(0).GetAs<oid_t>(),
+            table->GetOid(), database_oid,
+            static_cast<IndexType>(tup->GetValue(3).GetAs<oid_t>()),
+            static_cast<IndexConstraintType>(tup->GetValue(4).GetAs<oid_t>()),
+            tuple_schema, key_schema, key_attrs, tup->GetValue(4).GetAs<bool>());
+    std::shared_ptr<index::Index> index(
+            index::IndexFactory::GetIndex(index_metadata));
+    table->AddIndex(index);
+    tables_with_indexes.insert(table);
+    // Attributes must be changed once we have arraytype
+  }
+
+  // add tuples to index
+  for (storage::DataTable *table : tables_with_indexes) {
+    LOG_INFO("Install table_index");
+    auto schema = table->GetSchema();
+    size_t tg_count = table->GetTileGroupCount();
+    for (oid_t tg = 0; tg < tg_count; tg++) {
+      auto tile_group = table->GetTileGroup(tg);
+
+      for (int tuple_slot_id = START_OID; tuple_slot_id < DEFAULT_TUPLES_PER_TILEGROUP;
+           tuple_slot_id++) {
+        txn_id_t tuple_txn_id = tile_group->GetHeader()->GetTransactionId(tuple_slot_id);
+        if (tuple_txn_id != INVALID_TXN_ID) {
+          PELOTON_ASSERT(tuple_txn_id == INITIAL_TXN_ID);
+          std::unique_ptr<storage::Tuple> t(new storage::Tuple(schema, true));
+          for (size_t col = 0; col < schema->GetColumnCount(); col++) {
+            t->SetValue(col, tile_group->GetValue(tuple_slot_id, col), pool.get());
+          }
+          ItemPointer p(tile_group->GetTileGroupId(), tuple_slot_id);
+          auto txn = concurrency::TransactionManagerFactory::GetInstance()
+                  .BeginTransaction(IsolationLevelType::SERIALIZABLE);
+          ItemPointer *ip = nullptr;
+          table->InsertInIndexes(t.get(), p, txn, &ip);
+          tile_group->GetHeader()->SetIndirection(tuple_slot_id, ip);
+          concurrency::TransactionManagerFactory::GetInstance().CommitTransaction(
+                  txn);
+        }
+      }
+    }
+  }
+
+}
+
+void WalRecovery::ReplayAllTxns(){
+
+  for(auto it = commited_txns_.begin(); it!=commited_txns_.end(); it++){
+    ReplaySingleTxn(it->first);
+  }
+
+  // reset epoch
+  if (max_epoch_id_ != INVALID_EID) {
+    auto &epoch_manager = concurrency::EpochManagerFactory::GetInstance();
+    epoch_manager.Reset(max_epoch_id_+1);
+  }
+}
+
+
+void WalRecovery::ReplayLogFile(){
+
+  size_t curr_txn_offset = 0;
+
+  // Pass 1
+  ParseFromDisk(ReplayStage::PASS_1);
+
+
+  for(auto it = all_txns_.begin(); it!=all_txns_.end(); it++){
+
+    if(it->second.first != LogRecordType::TRANSACTION_COMMIT)
+      continue;
+
+    auto offset_pair = std::make_pair(curr_txn_offset, curr_txn_offset);
+    commited_txns_.insert(std::make_pair(it->first, offset_pair));
+
+    curr_txn_offset += it->second.second;
+  }
+
+  // Pass 2
+  log_buffer_  = new char[log_buffer_size_];
```

Comment:
I wonder if the recovery needs to read the log file twice from the disk. Since ```log_buffer_``` is constructed here after the first pass.
If this is the case, will it be better to populate the ```log_buffer_``` in the first phase and save the second disk read in the second phase?